File: python/dynamic_programming/UniquePathsII.py
class Solution:
    # @param obstacleGrid, a list of lists of integers
    # @return an integer
    def uniquePathsWithObstacles(self, obstacleGrid):
        m = len(obstacleGrid)
        n = len(obstacleGrid[0])
        
        opt = [[0 for i in xrange(n)] for i in xrange(m)]
        
        hasObstacle = False
        for i in xrange(m):
            if obstacleGrid[i][0] == 0 and not hasObstacle:
                opt[i][0] = 1
            else:
                opt[i][0] = 0
                hasObstacle = True
            
        hasObstacle = False
        for j in xrange(n):
            if obstacleGrid[0][j] == 0 and not hasObstacle:
                opt[0][j] = 1
            else:
                opt[0][j] = 0
                hasObstacle = True
        
        for i in xrange(1, m):
            for j in xrange(1, n):
                if obstacleGrid[i][j] == 1:
                    opt[i][j] = 0
                else:
                    opt[i][j] = opt[i-1][j] + opt[i][j-1]
        
        return opt[m-1][n-1]

if __name__ == "__main__":
    solution = Solution()

    print solution.uniquePathsWithObstacles([[0, 0], [0, 1]])

File: java/UniquePathsII.java
public class Solution {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        int row = obstacleGrid.length;
        if (row == 0) {
            return 0;
        }
        
        int col = obstacleGrid[0].length;
        int opt[][] = new int[row][col];
        if (obstacleGrid[0][0] == 1) {
            return 0;
        }
        
        opt[0][0] = 1;
        
        boolean hasObstacle = false;
        for (int i = 1; i < row; i++) {
            if (hasObstacle == false) {
                if (obstacleGrid[i][0] == 0) {
                    opt[i][0] = 1;
                } else {
                    hasObstacle = true;
                    opt[i][0] = 0;
                }
            } else {
                opt[i][0] = 0;
            }
        }
        
        hasObstacle = false;
        for (int i = 1; i < col; i++) {
            if (hasObstacle == false) {
                if (obstacleGrid[0][i] == 0) {
                    opt[0][i] = 1;
                } else {
                    hasObstacle = true;
                    opt[0][i] = 0;
                }
            } else {
                opt[0][i] = 0;
            }
        }
        
        for (int i = 1; i < row; i++) {
            for (int j = 1; j < col; j++) {
                if (obstacleGrid[i][j] == 1) {
                    opt[i][j] = 0;
                } else {
                    opt[i][j] = opt[i-1][j] + opt[i][j-1];
                }
            }
        }
        
        return opt[row-1][col-1];
    }
}
